torchtitan/datasets/experimental_datasets.py

```diff
@@ -0,0 +1,1212 @@
+        ]
+        assert len(countfiles) == 1
+        doc_counts = {}
+        pathsplit = (datapath, "")
+        while len(pathsplit[1]) == 0:
+            pathsplit = os.path.split(pathsplit[0])
+        pardir, dataset = pathsplit
+        self.dataset = dataset
+        with open(os.path.join(pardir, "meta", countfiles[0]), "r") as csvfile:
+            reader = csv.DictReader(csvfile)
+            for row in reader:
+                fullpath = row["dataset/filename"]
+                prefix = fullpath.find("/" + dataset) + 1
+                if prefix > 0:
+                    key = fullpath[prefix:]
+                    doc_counts[key] = int(row["documents"])
+
+        # Assemble document set owned by this worker:
+        # listdir, assemble shardfraglist (ind -> shard, frag)
+        shards = [
+            shard
+            for shard in os.listdir(datapath)
+            if os.path.isfile(os.path.join(datapath, shard))
+            and "arrow" in os.path.join(datapath, shard)
+        ]
+        shards.sort()  # Ensure consistent sharding across machines
+        start_frag = (rank * worldsize * len(shards)) // worldsize
+        end_frag = ((rank + 1) * worldsize * len(shards)) // worldsize
+        shardfrags = [
+            (shards[i // worldsize], i % worldsize) for i in range(start_frag, end_frag)
+        ]
+
+        # Read shardfrags, assemble doc list for each file shard (aggregating over fragments):
+        ndocs = -1
+        docset = {}  # shardid -> (min docid, max docid)
+        for i, (shard, frag) in enumerate(shardfrags):
+            ndocs = doc_counts[os.path.join(dataset, shard)]
+            self.docs_per_shard[shard] = ndocs
+            doc_start = (ndocs * frag) // worldsize
+            doc_end = (ndocs * frag + ndocs) // worldsize - 1  # Inclusive upper bound
+            if shard not in docset:
+                docset[shard] = [doc_start, doc_end]
+            min_d, max_d = docset[shard]
+            if doc_start < min_d:
+                docset[shard][0] = doc_start
+            if doc_end > max_d:
+                docset[shard][1] = doc_end
+
+        # Add all of this dataset's shard entries to self.docset
+        doccount = 0
+        for shardid in docset:
+            min_d = docset[shardid][0]
+            max_d = docset[shardid][1]
+            self.docset.append((shardid, min_d, max_d))
+            doccount += max_d - min_d + 1
+        self._len = doccount
+
+        if verbose:
+            logging.info(
+                f"    Worker {rank} ingested {len(shardfrags)} shard fragments from {dataset}"
+            )
+
+        # Shuffle shard files
+        if shuffle:
+            random.shuffle(self.docset)
+
+        self.docset_index = 0
+        self.chunk_index = -1
+
+        # Stats
+        self.epochs_seen = -1
+        self.tokens_seen = 0
+        self.docs_seen = 0
+        self.percent_seen = 0
+        self.lcg_state = seed + rank
+
+        self.state_params = [
```

Comment:
I see no reshard_params, so this dataset doesn't support resharding?

Reply:
Correct, this one particular dataset needs to sit under a `Scalable_Shard_Dataset` to support resharding